saleor/discount/models.py

```diff
@@ -103,7 +103,16 @@ def is_free(self):
         )
 
     def get_discount(self, channel: Channel):
-        voucher_channel_listing = self.channel_listings.filter(channel=channel).first()
+        """Return proper discount amount for given channel.
+
+        It operates over all channel listings as assuming that we have prefetched them.
+        """
+        voucher_channel_listing = None
+
+        for channel_listing in self.channel_listings.all():
+            if channel.id == channel_listing.channel_id:
+                voucher_channel_listing = channel_listing
```

Comment:
We could break loop if variable has been set.

Reply:
Breaking sounds good. Previously we had a filter here, but It was breaking prefetches which I applied and generates a lot of duplicated queries. I am creating a tasks, for better handling it, like passing channel_listing directly to this method.